tests/asynch/basic.rs

```diff
@@ -1,9 +1,103 @@
+use futures::try_join;
 use shuttle::sync::Mutex;
 use shuttle::{asynch, check_dfs, check_random, scheduler::PctScheduler, thread, Runner};
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 use test_env_log::test;
 
+async fn do_err(e: bool) -> Result<(), ()> {
+    if e {
+        Err(())
+    } else {
+        Ok(())
+    }
+}
+
+// Check that try_join on shuttle futures works as expected
+#[test]
+fn test_try_join() {
+    check_dfs(
+        || {
+            let f2 = do_err(true);
+            let f1 = do_err(false);
+            let res = asynch::block_on(async { try_join!(f1, f2) });
+            assert!(res.is_err());
+        },
+        None,
+    );
+}
+
+// Check that when we block on a future, it gets executed in both executions
+// (block_on task finishing first and main task finishing first).
+#[test]
+fn block_shuttle_future() {
+    let orderings_orig = Arc::new(AtomicUsize::new(0));
+    let async_accesses_orig = Arc::new(AtomicUsize::new(0));
+    let async_accesses = async_accesses_orig.clone();
+    let orderings = orderings_orig.clone();
+
+    check_dfs(
+        move || {
+            orderings.fetch_add(1, Ordering::SeqCst);
+            let async_accesses = async_accesses.clone();
+            asynch::block_on(async move {
+                async_accesses.fetch_add(1, Ordering::SeqCst);
+            });
+        },
+        None,
+    );
+
+    assert_eq!(2, orderings_orig.load(Ordering::SeqCst));
+    assert_eq!(2, async_accesses_orig.load(Ordering::SeqCst));
+}
+
+// Check that when we spawn and drop a future, it gets executed exactly one of two possible executions.
+#[test]
+fn drop_shuttle_future() {
+    let orderings_orig = Arc::new(AtomicUsize::new(0));
+    let async_accesses_orig = Arc::new(AtomicUsize::new(0));
+    let async_accesses = async_accesses_orig.clone();
+    let orderings = orderings_orig.clone();
+
+    check_dfs(
+        move || {
+            orderings.fetch_add(1, Ordering::SeqCst);
+            let async_accesses = async_accesses.clone();
+            asynch::spawn(async move {
+                async_accesses.fetch_add(1, Ordering::SeqCst);
+            });
+        },
+        None,
+    );
+
+    assert_eq!(2, orderings_orig.load(Ordering::SeqCst));
+    assert_eq!(1, async_accesses_orig.load(Ordering::SeqCst));
+}
+
+// Check that when we spawn and drop a future that immediately yields, it gets executed exactly one of two possible executions.
```

Comment:
maybe
```
// Same as `drop_shuttle_future`, but the inner task yields first, and might be cancelled part way through
``` 

and then it would be good to actually check that the "part way through" part is true -- there should be two executions that reach the `yield_now()`, but only one execution that reaches beyond there.

Reply:
Oh yeah, I like this idea.